Mathlib/GroupTheory/Submonoid/Units.lean

```diff
@@ -0,0 +1,322 @@
+the reduction from a submonoid to its unit group. -/
+@[to_additive " A Galois coinsertion exists between the coercion from a additive subgroup of
+additive units to a additive submonoid and the reduction from a additive submonoid to its unit
+group. " ]
+def ofUnits_units_gci : GaloisCoinsertion (Subgroup.ofUnits) (Submonoid.units (M := M)) :=
+  GaloisCoinsertion.monotoneIntro Submonoid.units_mono Subgroup.ofUnits_mono
+  Submonoid.ofunits_units_le Subgroup.ofUnits_units_eq
+
+@[to_additive]
+lemma ofUnits_units_gc : GaloisConnection (Subgroup.ofUnits) (Submonoid.units (M := M)) :=
+ofUnits_units_gci.gc
+
+namespace Submonoid
+
+section Units
+
+@[to_additive (attr := simp)]
+lemma mem_units_iff (S : Submonoid M) (x : Mˣ) : x ∈ S.units ↔
+    ((x : M) ∈ S ∧ ((x⁻¹ : Mˣ) : M) ∈ S) := Iff.rfl
+
+@[to_additive]
+lemma mem_units_of_coe_mem_coe_inv_mem (S : Submonoid M) {x : Mˣ} (h₁ : (x : M) ∈ S)
+    (h₂ : ((x⁻¹ : Mˣ) : M) ∈ S) : x ∈ S.units := ⟨h₁, h₂⟩
+
+@[to_additive]
+lemma coe_mem_of_mem_units (S : Submonoid M) {x : Mˣ} (h : x ∈ S.units) : (x : M) ∈ S := h.1
+
+@[to_additive]
+lemma coe_inv_mem_of_mem_units (S : Submonoid M) {x : Mˣ} (h : x ∈ S.units) :
+    ((x⁻¹ : Mˣ) : M) ∈ S := h.2
+
+@[to_additive]
+lemma coe_coe_inv_mul_coe_coe (S : Submonoid M) {x : Sˣ} :
+    ((x⁻¹ : Sˣ) : M) * ((x : Sˣ) : M) = 1 := congrArg ((↑) : S → M) (Units.inv_mul _)
+
+@[to_additive]
+lemma coe_coe_mul_coe_coe_inv (S : Submonoid M) {x : Sˣ} :
+    ((x : Sˣ) : M) * ((x⁻¹ : Sˣ) : M) = 1 := congrArg ((↑) : S → M) (Units.mul_inv _)
+
+@[to_additive]
+lemma mk_inv_mul_mk_eq_one (S : Submonoid M) {x : Mˣ} (h : x ∈ S.units) :
+    (⟨_, h.2⟩ : S) * ⟨_, h.1⟩ = 1 := Subtype.ext (Units.inv_mul _)
+
+@[to_additive]
+lemma mk_mul_mk_inv_eq_one (S : Submonoid M) {x : Mˣ} (h : x ∈ S.units) :
+    (⟨_, h.1⟩ : S) * ⟨_, h.2⟩ = 1 := Subtype.ext (Units.mul_inv _)
+
+@[to_additive]
+lemma mul_mem_units (S : Submonoid M) {x y : Mˣ} (h₁ : x ∈ S.units) (h₂ : y ∈ S.units):
+    x * y ∈ S.units := mul_mem h₁ h₂
+
+@[to_additive]
+lemma inv_mem_units (S : Submonoid M) {x : Mˣ} (h : x ∈ S.units) : x⁻¹ ∈ S.units := inv_mem h
+
+@[to_additive]
+lemma inv_mem_units_iff (S : Submonoid M) {x : Mˣ} : x⁻¹ ∈ S.units ↔ x ∈ S.units := inv_mem_iff
+
+/-- The equivalence between the subgroup of units of `S` and the type of units of `S`. -/
+@[to_additive (attr := simps!) " The equivalence between the additive subgroup of additive units of
+`S` and the type of additive units of `S`. "]
+def unitsEquivUnitsType (S : Submonoid M) : S.units ≃* Sˣ where
+  toFun := fun x => ⟨⟨_, x.2.1⟩, ⟨_, x.2.2⟩, S.mk_mul_mk_inv_eq_one x.2, S.mk_inv_mul_mk_eq_one x.2⟩
+  invFun := fun x => ⟨⟨_, _, S.coe_coe_mul_coe_coe_inv, S.coe_coe_inv_mul_coe_coe⟩, ⟨x.1.2, x.2.2⟩⟩
+  left_inv := fun _ => rfl
+  right_inv := fun _ => rfl
+  map_mul' := fun _ _ => rfl
+
+@[to_additive]
+lemma ge_ofUnits_iff_units_ge (S : Submonoid M) (H : Subgroup Mˣ) :
```

Comment:
I think the name should be `ofUnits_le_iff` or something similar.

Reply:
Righto, will look at this.